### DISCUSSION TITLE
Signing a PDF with CryptoID mToken.

### DISCUSSION THREAD

**Discussion by maxisimonazzi** [MatthiasValvekens/pyHanko#398] [Q&A]
Hello guys,  i have some CryptoID mTokens and i need to use them to sign pdf. is there any way to do it with pyhanko? or how can i "read" the certificates stored in the token?

**Comment by MatthiasValvekens** [comment]
> CryptoID mTokens

Explain what those are, please ;).

>  is there any way to do it with pyhanko?

If you have a way to sign arbitrary data: almost certainly yes. But the level of effort will depend on how you talk to those tokens and whether that mechanism is standard/already supported. If there's a PKCS#11 module for them: that'll do fine.

>  how can i "read" the certificates stored in the token?

That's one of the things you'll need to figure out. But the crucial part is not so much in reading the certificates, but in performing operations with the private key (which you typically can't just "read" out, that would defeat the purpose of the secure token).
